src/kokkos_abstraction.hpp

```diff
@@ -635,6 +638,38 @@ struct SpaceInstance<Kokkos::Cuda> {
 };
 #endif
 
+// Design from "Runtime Polymorphism in Kokkos Applications", SAND2019-0279PE
+template <typename ES = DevExecSpace, typename MS = DevMemSpace>
+struct DeviceDeleter {
+  template <typename T>
+  void operator()(T *ptr) {
+    Kokkos::kokkos_free<MS>(ptr);
+  }
+};
+
+template <typename T, typename ES = DevExecSpace, typename MS = DevMemSpace>
+std::unique_ptr<T, DeviceDeleter<ES, MS>> DeviceAllocate() {
+  PARTHENON_REQUIRE(std::is_trivially_destructible<T>::value,
+                    "DeviceAllocate only supports trivially destructible classes!");
+  auto *p = static_cast<T *>(Kokkos::kokkos_malloc<MS>(sizeof(T)));
```

Comment:
_Technically_ you should put this pointer (and in `DeviceCopy`) into `unique_ptr` before calling any of the Kokkos functions, since they may throw. If you don't, then this would leak should you catch the exception and continue running. However, this is a pretty minor issue - exceptions almost always result in code termination in parthenon.

Reply:
That's a good point, this is a pretty generic function so it's best to do things right here. Done.